sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/UnboundedRabbitMqReader.java

```diff
@@ -0,0 +1,173 @@
+
+class UnboundedRabbitMqReader extends UnboundedSource.UnboundedReader<RabbitMqMessage> {
+
+  private final RabbitMQSource source;
+
+  private TimestampPolicy timestampPolicy;
+  private RecordIdPolicy recordIdPolicy;
+  private TimestampPolicyContext context;
+  private RabbitMqMessage currentRecord;
+  private String queueName;
+  private final RabbitMQCheckpointMark checkpointMark;
+
+  UnboundedRabbitMqReader(RabbitMQSource source, RabbitMQCheckpointMark checkpointMark) {
+    this.source = source;
+    this.currentRecord = null;
+    this.recordIdPolicy = source.spec.recordIdPolicy();
+    this.context = new TimestampPolicyContext(true, BoundedWindow.TIMESTAMP_MIN_VALUE);
+    this.checkpointMark =
+        checkpointMark != null ? checkpointMark : new RabbitMQCheckpointMark(source.spec.uri());
+    mkTimestampPolicy(
+        Optional.ofNullable(checkpointMark).map(RabbitMQCheckpointMark::getWatermark));
+  }
+
+  private TimestampPolicy mkTimestampPolicy(Optional<Instant> prevWatermark) {
+    if (this.timestampPolicy == null) {
+      this.timestampPolicy =
+          source.spec.timestampPolicyFactory().createTimestampPolicy(prevWatermark);
+    }
+    return this.timestampPolicy;
+  }
+
+  @Override
+  public Instant getWatermark() {
+    return timestampPolicy.getWatermark(context, currentRecord);
+  }
+
+  @Override
+  public UnboundedSource.CheckpointMark getCheckpointMark() {
+    return checkpointMark;
+  }
+
+  @Override
+  public RabbitMQSource getCurrentSource() {
+    return source;
+  }
+
+  @Override
+  public byte[] getCurrentRecordId() {
+    return Optional.ofNullable(currentRecord).map(recordIdPolicy).get();
+  }
+
+  @Override
+  public Instant getCurrentTimestamp() {
+    return Optional.ofNullable(currentRecord).map(timestampPolicy::getTimestampForRecord).get();
+  }
+
+  @Override
+  public RabbitMqMessage getCurrent() {
+    if (currentRecord == null) {
+      throw new NoSuchElementException();
+    }
+    return currentRecord;
+  }
+
+  @Override
+  public boolean start() throws IOException {
+    try {
+      Channel channel = checkpointMark.getChannel();
+      checkpointMark.startReading();
+
+      // TODO: rework all of this
+      queueName = source.spec.queue();
+      if (source.spec.queueDeclare()) {
+        // declare the queue (if not done by another application)
+        // channel.queueDeclare(queueName, durable, exclusive, autoDelete, arguments);
+        channel.queueDeclare(queueName, false, false, false, null);
+      }
+      if (source.spec.exchange() != null) {
+        if (source.spec.exchangeDeclare()) {
+          channel.exchangeDeclare(source.spec.exchange(), source.spec.exchangeType());
+        }
+        if (queueName == null) {
+          queueName = channel.queueDeclare().getQueue();
+        }
+        channel.queueBind(queueName, source.spec.exchange(), source.spec.routingKey());
+      }
+    } catch (IOException e) {
+      checkpointMark.stopReading();
```

Comment:
Notable changes; lets the checkpoint mark know it can close its underlying connection if nothing is reading from it and `finalizeCheckpoint` is called